File: examples/java/Case07.java
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.SynchronousQueue;
import java.util.concurrent.atomic.AtomicBoolean;

final class Case07 {
    static final class CancelSignal {
        private final AtomicBoolean cancelled = new AtomicBoolean();
        private final CopyOnWriteArrayList<Runnable> listeners = new CopyOnWriteArrayList<>();
        boolean isCancelled() { return cancelled.get(); }
        void cancel() {
            if (cancelled.compareAndSet(false, true)) for (Runnable listener : listeners) listener.run();
        }
        Runnable onCancel(Runnable listener) {
            listeners.add(listener);
            if (isCancelled()) listener.run();
            return () -> listeners.remove(listener);
        }
    }

    private static int processBounded(int item) { return item * 2; }

    // faultscope:begin fs-c07.blocking-send-without-cancel
    static void blockingSendWithoutCancel(SynchronousQueue<Integer> out, int result,
                                          CountDownLatch entered) throws InterruptedException {
        entered.countDown();
        out.put(result); // No receiver: the cancel token cannot release this wait.
    }
    // faultscope:end fs-c07.blocking-send-without-cancel

    // faultscope:begin fs-c07.precheck-then-block
    static void precheckThenBlock(CancelSignal cancel, SynchronousQueue<Integer> out,
                                  int item, CountDownLatch entered) throws InterruptedException {
        if (cancel.isCancelled()) return;
        int result = processBounded(item);
        blockingSendWithoutCancel(out, result, entered);
    }
    // faultscope:end fs-c07.precheck-then-block

    // faultscope:begin fs-c07.blocking-send-with-cancel
    static boolean blockingSendWithCancel(CancelSignal cancel, SynchronousQueue<Integer> out,
                                          int result, CountDownLatch entered) throws InterruptedException {
        if (cancel.isCancelled()) return false;
        Runnable unregister = cancel.onCancel(Thread.currentThread()::interrupt);
        try {
            entered.countDown();
            out.put(result); // Cancellation interrupts this blocking operation.
            return true;
        } catch (InterruptedException interrupted) {
            if (cancel.isCancelled()) return false;
            throw interrupted;
        } finally {
            unregister.run();
        }
    }
    // faultscope:end fs-c07.blocking-send-with-cancel

    // faultscope:begin fs-c07.long-work-check-cancel
    static boolean longWorkCheckCancel(CancelSignal cancel, int units) {
        for (int index = 0; index < units; index++) {
            if (index % 1024 == 0 && cancel.isCancelled()) return false;
            int boundedUnit = index * index; // One bounded unit needs no check per instruction.
        }
        return true;
    }
    // faultscope:end fs-c07.long-work-check-cancel

    // faultscope:begin fs-c07.downstream-without-cancellation
    static void downstreamWithoutCancellation(CountDownLatch release, CountDownLatch entered)
            throws InterruptedException {
        entered.countDown();
        release.await(); // This API has no cancel token.
    }
    // faultscope:end fs-c07.downstream-without-cancellation
}

File: examples/go/case07.go
package examplesgo

import "context"

func case07ProcessBounded(item int) int { return item * 2 }

// faultscope:begin fs-c07.blocking-send-without-cancel
func BlockingSendWithoutCancel(out chan<- int, result int, entered chan<- struct{}) {
	close(entered)
	out <- result // With no receiver, cancellation cannot release this send.
}

// faultscope:end fs-c07.blocking-send-without-cancel

// faultscope:begin fs-c07.precheck-then-block
func PrecheckThenBlock(ctx context.Context, out chan<- int, item int, entered chan<- struct{}) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	result := case07ProcessBounded(item)
	BlockingSendWithoutCancel(out, result, entered)
	return nil
}

// faultscope:end fs-c07.precheck-then-block

// faultscope:begin fs-c07.blocking-send-with-cancel
func BlockingSendWithCancel(ctx context.Context, out chan<- int, result int, entered chan<- struct{}) error {
	close(entered)
	select {
	case out <- result:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}

// faultscope:end fs-c07.blocking-send-with-cancel

// faultscope:begin fs-c07.long-work-check-cancel
func LongWorkCheckCancel(ctx context.Context, units int) error {
	for i := 0; i < units; i++ {
		if i%1024 == 0 {
			if err := ctx.Err(); err != nil {
				return err
			}
		}
		_ = i * i // One bounded unit need not check after every instruction.
	}
	return nil
}

// faultscope:end fs-c07.long-work-check-cancel

// faultscope:begin fs-c07.downstream-without-cancellation
func DownstreamWithoutCancellation(release <-chan struct{}, entered chan<- struct{}) {
	close(entered)
	<-release // This API has no cancellation input.
}

// faultscope:end fs-c07.downstream-without-cancellation
